### DISCUSSION TITLE
Plug and play regularisers in pfb

### DISCUSSION THREAD

**Discussion by landmanbester** [ratt-ru/pfb-clean#101] [Ideas]
Chatting to @TariqBlecher about incorporating some learnt regularisers into one of the minor cycles in pfb. Suggest we start with the approach in https://arxiv.org/pdf/2312.00125.pdf. I have created a new `taivas` worker (feel free to change the name @TariqBlecher) to get the ball rolling. After running the `init` and `grid` steps using eg.

```
$ pfb init --ms path/to/MS -o out/test -cpi -1 -ipi -1 -dc DATA -fov 2.0 --scheduler distributed -nw 4 -ntd 2 -nvt 2
$ pfb grid -o out/test -nb 1 -fov 1.5 -srf 2 --robustness -1.5 --scheduler distributed -nw 4 -ntd 2 -nvt 2
```

You should be able to run the `taivas` worker with

```
$ pfb taivas -o out/test -nb 1 --niter 5 --scheduler threads -nw 1 -ntd 1 -nvt 8
```

This will drop you into ipdb (you may have to `pip install ipdb`) at the point where the new minor cycle can be inserted. Don't worry about the parallelisation settings (i.e. the stuff after `--scheduler` for the time being, the above should ok to start with).
 
My suggestion is that you start by:

1. Defining and testing a psf convolution function that can run on pytorch. You can test it against the cpu version `psf_concolve(.)` that is available at the debug point.
2.  Create the likelihood function `vdot(x, psf_convolve(x) - 2*dirty)` for pytorch. I assume this will have to support autodiff (at least grad). Again you can test it against the numpy implementation (the grad is simply `2*psf_convolve(x) - 2*dirty` so you can test that too).
3. Load a pretrained model and run crrfista from the repo to get a new model. 

I have set `-nb 1 ` above assuming you are going to start with a single band. If you want to slice out a subset of the frequency axis you can just use `--freq-range fmin:fmax` with fmin and fmax in Hz when you run the `init` worker. Going to multiple bands, I think a sensible strategy is to run them separately in parallel for now. We can think about imposing regularisation along the frequency axis at a later stage. My thinking is that we can impose smoothness along frequency during the forward step, thus leaving the rest of the algo essentially unchanged. 

Let's discuss progress/problems etc. here

**Comment by landmanbester** [comment]
@TariqBlecher @y-mhiri I finally got around to refactoring the new deconv modules. I haven't tested them but I think they should at least run. `pfb-imaging` now follows the [`hip-cargo`](https://github.com/landmanbester/hip-cargo) convention. I believe the structure of the project is easier to follow now (all imports should now live at the top of files and no more dodgy **kw unpacking). The idea is for the files in src/pfb_imaging/cli to define the CLI's for the different workers (lightweight imports live here). You can run 
```
pfb --help
```
to see a list of them as usual. Cabs are auto-generated if you install the pre-commit hooks. The core implementations live in the `src/pfb_imaging/core` module as plain python functions. Place all your heavy imports at the top of these files. I've not gotten to the generalised deconvolution structure yet but I think we should aim for general optimisation routines like `forward-backward` or `primal-dual` to live in the `src/pfb_imaging/opt` module. Specific implementations of them (i.e. those for which we specify the _grad and _prox methods) should live in the `src/pfb_imaging/deconv` module. I tried to restructure but I might have missed something. At some point I'll try to refactor `sara` and `ista` accordingly. Let's discuss any questions you might have here

**Comment by y-mhiri** [comment]
Hi @landmanbester  @TariqBlecher ,
I've been able to install easily on a new environment. I tested the init, grid and ista worker on the multi spectral band SNRG55 dataset. 
The CLI is in fact way more responsive ! 

Everything mostly ok, only a few typos : 
- An import error in `deconv.ista_fb` and `core.airi`
- A few unexpected argument error in the `cli.ista` and  `cli.airi` wrapper
- Not sure why but last log on number threads used by `numba` gives a ValueError (`"Threading layer is not initialized."`) even if I actually used 24 threads on the run.
I've made a commit on `prox_op_Tariq` branch, correcting those typos.

On AIRI, missing torch dependencies specifications, it might be nice to add to the `pytoml` an `airi` profil on the optional-dependencies section. Also, @TariqBlecher do you have a link to download a set of pre-trained denoisers ? 

Free to set up a meeting in the next weeks to see where to go next.

**Comment by landmanbester** [comment]
Thanks @y-mhiri for the fixes. 

> Not sure why but last log on number threads used by numba gives a ValueError ("Threading layer is not initialized.") even if I actually used 24 threads on the run.

Does numba actually get used in `ista`? If so then this is probably a bug.

 > On AIRI, missing torch dependencies specifications, it might be nice to add to the pytoml an airi profil on the optional-dependencies section.

Yes, I'll add these as optional dependencies.

> Also, @TariqBlecher do you have a link to download a set of pre-trained denoisers ?

Please do share these if you have them @TariqBlecher.

Let's set up a meeting once I have the `typer` branch merged. This has been a way bigger undertaking than I anticipated but I should be done fairly soon
